### DISCUSSION TITLE
How to serde i64 with string by default?

### DISCUSSION THREAD

**Discussion by lz1998** [tokio-rs/axum#576] [Q&A] [answered]
I want to serde i64 with string, but my struct is generated by prost, so I can't add `#[serde(with = "string")]`.

**Comment by davidpdrsn** [comment, marked as answer]
Is this a question about axum or prost?

Regardless I suppose you can use https://docs.rs/prost-build/latest/prost_build/struct.Config.html#method.field_attribute or https://docs.rs/prost-build/latest/prost_build/struct.Config.html#method.type_attribute to add the attributes on the prost side.

**Reply by lz1998** [reply]
I have tried this, but it doesn't work, so I am trying to find some other methods to serde i64 with string by default. Maybe some axum serde global config.
```rust
    let mut cfg = prost_build::Config::new();
    cfg.type_attribute(".", "#[derive(serde::Serialize, serde::Deserialize)] #[serde(default)]");
    cfg.type_attribute("int64", "#[serde(with = \"string\")]");
    cfg.type_attribute(".int64", "#[serde(with = \"string\")]");
    cfg.type_attribute("i64", "#[serde(with = \"string\")]");
    cfg.type_attribute(".i64", "#[serde(with = \"string\")]");
```

**Reply by davidpdrsn** [reply]
I think you need to ask the prost people. I'm not familiar enough with prost to know.

Nothing axum specific should be necessary, and in fact can't override prost/serde stuff at all actually.

**Reply by lz1998** [reply]
If the struct is not generated by prost. Can I set a custom formatter for serde_json in axum?

https://github.com/serde-rs/json/issues/833

**Reply by davidpdrsn** [reply]
> If the struct is not generated by prost.

I mean you can do whatever you want 😊 axum doesn't care but also doesn't provide anything for you since its not specific to axum in any way.

> Can I set a custom formatter for serde_json in axum?

I don't understand what you mean. Whats a formatter in the context of serde_json? You can write your own version of [`Json`](https://docs.rs/axum/0.3.4/axum/struct.Json.html) that implements [`IntoResponse`](https://docs.rs/axum/0.3.4/axum/response/trait.IntoResponse.html) makes the JSON output look different if thats what you mean.
